### PR TITLE
feat: Elasticsearch Metadata Extractors

### PR DESCRIPTION
**Pull request by mgorsk1**
### Summary of Changes

Introduces extractors for Elasticsearch metadata:
- Index extractor utilizing TableMetadata to feed Amundsen index metadata from ES
- Column Stats extractor utilizing ES Aggregation to provide stats on numerical properties

### Tests

- added tests for each extractor

### Documentation

- added docs for each extractor

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

**Review by feng-tao — COMMENTED** [review]
hey @mgorsk1 , I maybe missing something fundamental, but ES shouldn't be the source of truth as the index is always rebuilt from the metadata store. Why we want to extract from ES instead of metadata?

**Comment by mgorsk1** [comment]
Do you mean that ES Indexes are always populated from some other DBs like postgres or snowflake and we should only ingest metadata from those systems ?

I can think of several use cases where ES index is the only source of truth (like in logging use-cases, ingesting OCR data). On the other hand, if the index indeed was created by extracting table from (for example) postgresql then it's also valuable to have this information for the lineage purposes.

**Comment by feng-tao** [comment]
> you mean that ES Indexes are always populated from some other DBs like postgres or snowflake and we should only ingest metadata from those systems ?

yeah, what I said is only from Amundsen context that ES is always used for search purpose but not for metadata source of truth. ES certainly could be source of truth for other use cases like you said.

**Comment by mgorsk1** [comment]
Ok, maybe it's worth to update the docs though if that was unclear. The purpose of the extractor is indeed for users relying on ES for something else than Amundsen (like use-cases I mentioned above).

File: databuilder/databuilder/extractor/es_base_extractor.py
Comment (feng-tao) [inline databuilder/databuilder/extractor/es_base_extractor.py:19]: could you give a comment on config value ?

**Review by feng-tao — COMMENTED** [review]
Also if the extractor is using for the metadata stored in ES, have you looked into es dbapi (e.g https://github.com/preset-io/elasticsearch-dbapi ) given we have a dbapi extractor?

File: databuilder/databuilder/extractor/es_base_extractor.py
Comment (mgorsk1) [reply databuilder/databuilder/extractor/es_base_extractor.py:19]: it's included it in docs.

Comment (mgorsk1) [reply databuilder/databuilder/extractor/es_base_extractor.py:19]: As for dbapi - from the docs I can see that `This library supports Elasticsearch 7.X versions.` - I tried it with 6.8.13 and failed. Also - I am not sure the column statistics part would be achieved using this library.

**Comment by mgorsk1** [comment]
ping @feng-tao for re-review

**Comment by feng-tao** [comment]
@mgorsk1 sorry for the delay, but I kinda packed this week and may not have time to go through the code. Maybe other folks (cc @verdan @allisonsuarez ) to take a look...

**Comment by verdan** [comment]
Sure thing. Will be able to review tomorrow.

File: databuilder/README.md
Comment (verdan) [inline databuilder/README.md:1310]: any reason we are using TableMetadata model here? Shouldn't this be a new model for this use case?

File: databuilder/README.md
Comment (verdan) [inline databuilder/README.md:1350]:
```suggestion
    f'extractor.es_metadata.{ElasticsearchMetadataExtractor.CLUSTER}': 'demo',
    f'extractor.es_metadata.{ElasticsearchMetadataExtractor.SCHEMA}': 'dev',
    f'extractor.es_metadata.{ElasticsearchMetadataExtractor.ELASTICSEARCH_CLIENT_CONFIG_KEY}': Elasticsearch()
```

File: databuilder/README.md
Comment (verdan) [inline databuilder/README.md:1403]:
```suggestion
    f'extractor.es_metadata.{ElasticsearchColumnStatsExtractor.CLUSTER}': 'demo',
    f'extractor.es_metadata.{ElasticsearchColumnStatsExtractor.SCHEMA}': 'dev',
    f'extractor.es_metadata.{ElasticsearchColumnStatsExtractor.ELASTICSEARCH_CLIENT_CONFIG_KEY}': Elasticsearch()
```

**Review by verdan — CHANGES_REQUESTED** [review]

File: databuilder/README.md
Comment (mgorsk1) [reply databuilder/README.md:1310]: Hm I think TableMetadata is a very diverse concept and can be applied on NoSQL DBs as well. Unless we would want a different type of resource in UI for those but I am not sure how would it differ from regular table view and whether it makes sense to make such distinction.

**Review by verdan — APPROVED** [review]
LGTM

**Comment by mgorsk1** [comment]
@feng-tao need an ok from codeowner 🙏

**Review by feng-tao — APPROVED** [review]

**Pull request merged**
